keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+#### Huge increase of log volume
+
+This effort should result in reducing costs of log storage and analysis in production setups. Introducing structured logs will allow for creating better indices and reducing the size of data needed for analysis. From log throughput perspective there will be an increase of log size causing a pressure of log ingestion (disk, logging agents, logging API). To reduce the potential impact we will consider adding additional configuration flags allowing user to specify which metadata should be dropped
+
+## Design Details
+
+### Mapping between logging methods
+
+Following table provides a mapping between klog and logr methods to be used during the automated migration.
+
+| **klog logger** | **logr logger**
+| - |-  |
+`klog.V(level).Info*(format/args[0], args)` | `<module>.Logger.V(level+2).Info(format/args[0], “args”, args)`
+`klog.Info*(format/args[0], args)` | `<module>.Logger.V(2).Info(format/args[0], “args”, args)`
+`klog.Warning*(format/args[0], args)` | `<module>.Logger.V(1).Info(format/args[0], “args”, args)`
+`klog.Error*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args)` or after manual verification that this is not a proper error log: `<module>.Logger.Info(format/args[0], “error”, error(format/args[0]), “args”, args)`
+`klog.Fatal*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args, “stacks”, k8slog.AllStacks()) os.Exit(255)`
+`klog.Exit*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args) os.Exit(1)`
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+#### Alpha
+
+* Introduce a flag allowing to switch to new log format
+* Mechanism for defining standard key fields is implemented
+* `klog` is deprecated and should not be used for new logs
+* Automatic stage of migration is finished 
+
+#### Beta
+
+* Structured logs were manually verified 
+* New Logging guide for K8s is available
+
+#### GA
+ 
+* Structured logs are the default
+* Instructions on migrating other kubernetes repositories
+
+#### Deprecating old logging format
+
+* After 2 k8s versions old logging format removed 
+
+### Upgrade / Downgrade Strategy
+
+Double writing during migration will ensure identical logs without any configuration changes. Switching to new format will require passing a flag to each component or waiting for full removal of klog.
+
+### Version Skew Strategy
+
+During migration not all components will be able to support the new log format. It should be for users to decide if they want to opt into new logging format on partial set of components or wait for full migration.
+
+## Implementation History
+
+TBD
+
+## Drawbacks
+
+## Alternatives
```

Comment:
As somebody with no background into this while reading the document throughout I had one question in my mind: why not tracing?
OpenTelemetry and the likes would allow for sampling( which would ease the increased volume of logs + storage costs), and alleviate the "_ok, look for the log with this data/ id and see what it is related to_".

If the maturity level of the OpenTracing/ OpenCensus/ OpenTelemetry is the issue I'd still like to see that mentioned here.

Reply:
Tracing similarly to monitoring is a complementary feature to logging and not the substitution for it. Depending on your use case you might want to use one of those and not the other with monitoring providing the most aggregated information, tracing being more specific and logging giving you access to the most row data.
So in my mind there is no question why one and not the other because in the end it is good to have all of those. There is however a question which should be implemented first. As this change will enable much easier implementation of tracing for Kubernetes in the future by introduction of context propagation which is needed by OpenTracing/OpenCensus/OpenTelemetry I believe that it should be implemented first.

Reply:
Hm... interesting. I see tracing as — and give me some rope here — ordered structured logs so in my mind there’s not a need for both tracing and logs. 

Getting from a structured log to a trace in my mind is just adding a couple fields: `trace_id` and `parent_id`. The service name, the timestamps, the helpful context data are all already there.

Reply:
I have a bit different model in my mind when it comes to tracing.
Most modern tracing systems like zipkin or jaeger when it comes to traces operate on concepts like spans and traces. Spans being abstract concepts which have a name, some beginning and end timestamp and a list of tags. In jaeger spans can have logs associated to them but in general in my mind those are independent so that for instance you could have spans without any logs associated to them and you could have logs which does not really fit well into trace/span model like for instance logs from long running batch processing process.